packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -157,84 +191,123 @@ class PerformanceOverlayLayer extends Layer {
+    super.detach();
+    Layer child = firstChild;
+    while (child != null) {
+      child.detach();
+      child = child.nextSibling;
+    }
+  }
+
   /// Adds the given layer to the end of this layer's child list
   void append(Layer child) {
     assert(child != this);
-    assert(child != _firstChild);
-    assert(child != _lastChild);
-    assert(child._parent == null);
-    assert(child._nextSibling == null);
-    assert(child._previousSibling == null);
+    assert(child != firstChild);
+    assert(child != lastChild);
+    assert(child.parent == null);
+    assert(!child.attached);
+    assert(child.nextSibling == null);
+    assert(child.previousSibling == null);
     assert(() {
       Layer node = this;
       while (node.parent != null)
         node = node.parent;
       assert(node != child); // indicates we are about to create a cycle
       return true;
     });
-    child._parent = this;
-    child._previousSibling = _lastChild;
-    if (_lastChild != null)
-      _lastChild._nextSibling = child;
+    adoptChild(child);
+    child._previousSibling = lastChild;
+    if (lastChild != null)
+      lastChild._nextSibling = child;
     _lastChild = child;
     _firstChild ??= child;
+    assert(child.attached == attached);
   }
 
-  void _remove(Layer child) {
-    assert(child._parent == this);
-    assert(_debugUltimatePreviousSiblingOf(child, equals: _firstChild));
-    assert(_debugUltimateNextSiblingOf(child, equals: _lastChild));
+  /// Implementation of [Layer.remove].
+  void _removeChild(Layer child) {
+    assert(child.parent == this);
+    assert(child.attached == attached);
+    assert(_debugUltimatePreviousSiblingOf(child, equals: firstChild));
+    assert(_debugUltimateNextSiblingOf(child, equals: lastChild));
     if (child._previousSibling == null) {
       assert(_firstChild == child);
       _firstChild = child._nextSibling;
     } else {
-      child._previousSibling._nextSibling = child._nextSibling;
+      child._previousSibling._nextSibling = child.nextSibling;
     }
     if (child._nextSibling == null) {
-      assert(_lastChild == child);
-      _lastChild = child._previousSibling;
+      assert(lastChild == child);
+      _lastChild = child.previousSibling;
     } else {
-      child._nextSibling._previousSibling = child._previousSibling;
+      child.nextSibling._previousSibling = child.previousSibling;
     }
+    assert((firstChild == null) == (lastChild == null));
+    assert(firstChild == null || firstChild.attached == attached);
+    assert(lastChild == null || lastChild.attached == attached);
+    assert(firstChild == null || _debugUltimateNextSiblingOf(firstChild, equals: lastChild));
+    assert(lastChild == null || _debugUltimatePreviousSiblingOf(lastChild, equals: firstChild));
     child._previousSibling = null;
     child._nextSibling = null;
-    child._parent = null;
+    dropChild(child);
+    assert(!child.attached);
   }
 
   /// Removes all of this layer's children from its child list
```

Comment:
It might be worth pointing out that dropChild() is applied to about-to-be orphan.